### ISSUE TITLE
`raw_exec` plugin crashing with segfault

### ISSUE THREAD

**Issue by dvusboy** [hashicorp/nomad#1136]
### Nomad version
Output from `nomad version`:
```
Nomad v0.3.2
```
I am using the pre-built binary from `releases.hashicorp.com`

### Operating system and Environment details
```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

### Issue
`raw_exec` plugin is crashing with a signal 11 when I'm running a shell script that eventually tries to launch `docker-compose`. However, I don't believe the shell-script, specified as `artifact_source`, is retrieved because I don't see it in the `alloc` directory.

### Reproduction steps
This would be very involved, as the docker images are hosted in a private registry. That said, I'm providing snippets from the `systemd` journal and many of the artifacts such as the compose file, shell-script and the JSON job spec in [this gist](https://gist.github.com/dvusboy/db875543f3bd83e57b079cbcb9c02e81)

### Nomad Server logs (if appropriate)
I am running a 3-member server cluster. But this is the output of the `alloc-status`:

```
$ nomad alloc-status 82684350
ID            = 82684350
Eval ID       = 248a9d2f
Name          = gromacs-trimethylamine.trimethylamine-vdw-17_group[0]
Node ID       = 04294207
Job ID        = gromacs-trimethylamine
Client Status = failed

==> Task Resources
Task: "trimethylamine-vdw-17"
CPU    Memory MB  Disk MB  IOPS  Addresses
12000  2000       190      0

==> Task "trimethylamine-vdw-17" is "dead"
Recent Events:
Time                   Type                   Description
28/04/16 18:38:49 PDT  Restarts Exceeded      <none>
28/04/16 18:38:49 PDT  Driver Failure         error starting process via the plugin: unexpected EOF
28/04/16 18:38:49 PDT  Downloading Artifacts  Client is downloading artifacts
28/04/16 18:38:49 PDT  Received               Task received by client
```

### Nomad Client logs (if appropriate)
See the gist referenced above.

### Job file (if appropriate)
See the gist.

**Comment by dvusboy** [comment]
Here's a even simpler `raw_exec` run:

#### job spec:
```
job "test-whoami" {
	datacenters = ["rnd"]
	type = "batch"
	group "grp" {
		count = 1
		task "whoami" {
			driver = "raw_exec"
			config {
				command = "/usr/bin/whoami"
			}
			resources {
				cpu = 100
				memory = 10
			}
		}
	}
}
```

#### `systemd` journal
```
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] client: starting task runners for alloc '1003f732-e08d-7091-43a0-eda0c56bbc5d'
Apr 29 14:38:26 <host> nomad[140535]: client: starting task runners for alloc '1003f732-e08d-7091-43a0-eda0c56bbc5d'
Apr 29 14:38:26 <host> nomad[140535]: client: starting task context for 'whoami' (alloc '1003f732-e08d-7091-43a0-eda0c56bbc5d')
Apr 29 14:38:26 <host> nomad[140535]: plugin: starting plugin: /usr/bin/nomad []string{"/usr/bin/nomad", "executor", "/var/lib/nomad/alloc/1003f732-e08d-7091-43a0-eda0c56bbc5d/whoami/whoami-executor.out"}
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] client: starting task context for 'whoami' (alloc '1003f732-e08d-7091-43a0-eda0c56bbc5d')
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: starting plugin: /usr/bin/nomad []string{"/usr/bin/nomad", "executor", "/var/lib/nomad/alloc/1003f732-e08d-7091-43a0-eda0c56bbc5d/whoami/whoami-executor.out"}
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: waiting for RPC address for: /usr/bin/nomad
Apr 29 14:38:26 <host> nomad[140535]: plugin: waiting for RPC address for: /usr/bin/nomad
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: 2016/04/29 14:38:26 [DEBUG] plugin: plugin address: unix /tmp/plugin683611674
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: 2016/04/29 14:38:26 [DEBUG] plugin: plugin address: unix /tmp/plugin683611674
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: panic: runtime error: invalid memory address or nil pointer dereference
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: panic: runtime error: invalid memory address or nil pointer dereference
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: [signal 0xb code=0x1 addr=0x0 pc=0x92dcf8]
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad:
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: goroutine 26 [running]:
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: panic(0x1124d20, 0xc82000e070)
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad:         /opt/go/src/runtime/panic.go:464 +0x3e6
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: github.com/hashicorp/nomad/client/driver/executor.(*UniversalExecutor).configureTaskDir(0xc8202b2000, 0x0, 0x0)
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad:         /opt/gopath/src/github.com/hashicorp/nomad/client/driver/executor/executor.go:451 +0x68
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: github.com/hashicorp/nomad/client/driver/executor.(*UniversalExecutor).LaunchCmd(0xc8202b2000, 0xc820166320, 0xc820166370, 0x8, 0x0, 0x0)
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad:         /opt/gopath/src/github.com/hashicorp/nomad/client/driver/executor/executor.go:190 +0x27f
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: github.com/hashicorp/nomad/client/driver.(*ExecutorRPCServer).LaunchCmd(0xc8201731e0, 0xc820166320, 0xc820166370, 0xc8202f4100, 0x0, 0x0)
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad:         /opt/gopath/src/github.com/hashicorp/nomad/client/driver/executor_plugin.go:97 +0x64
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: reflect.Value.call(0x104d900, 0x1286e98, 0x13, 0x1306748, 0x4, 0xc82007fee8, 0x3, 0x3, 0x0, 0x0, ...)
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad:         /opt/go/src/reflect/value.go:435 +0x120d
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: reflect.Value.Call(0x104d900, 0x1286e98, 0x13, 0xc82007fee8, 0x3, 0x3, 0x0, 0x0, 0x0)
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad:         /opt/go/src/reflect/value.go:303 +0xb1
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: net/rpc.(*service).call(0xc820174dc0, 0xc820174d80, 0xc820164980, 0xc82024c480, 0xc8201734e0, 0x113c640, 0xc820164c50, 0x199, 0xe8d700, 0xc8202f4100, ...)
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad:         /opt/go/src/net/rpc/server.go:383 +0x1c2
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad: created by net/rpc.(*Server).ServeCodec
Apr 29 14:38:26 <host> nomad[140535]: plugin: nomad:         /opt/go/src/net/rpc/server.go:477 +0x49d
Apr 29 14:38:26 <host> nomad[140535]: plugin: /usr/bin/nomad: plugin process exited
Apr 29 14:38:26 <host> nomad[140535]: client: failed to start task 'whoami' for alloc '1003f732-e08d-7091-43a0-eda0c56bbc5d': error starting process via the plugin: unexpected EOF
Apr 29 14:38:26 <host> nomad[140535]: client: Not restarting task: whoami for alloc: 1003f732-e08d-7091-43a0-eda0c56bbc5d
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: [signal 0xb code=0x1 addr=0x0 pc=0x92dcf8]
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad:
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: goroutine 26 [running]:
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: panic(0x1124d20, 0xc82000e070)
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad:         /opt/go/src/runtime/panic.go:464 +0x3e6
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: github.com/hashicorp/nomad/client/driver/executor.(*UniversalExecutor).configureTaskDir(0xc8202b2000, 0x0, 0x0)
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad:         /opt/gopath/src/github.com/hashicorp/nomad/client/driver/executor/executor.go:451 +0x68
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: github.com/hashicorp/nomad/client/driver/executor.(*UniversalExecutor).LaunchCmd(0xc8202b2000, 0xc820166320, 0xc820166370, 0x8, 0x0, 0x0)
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad:         /opt/gopath/src/github.com/hashicorp/nomad/client/driver/executor/executor.go:190 +0x27f
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: github.com/hashicorp/nomad/client/driver.(*ExecutorRPCServer).LaunchCmd(0xc8201731e0, 0xc820166320, 0xc820166370, 0xc8202f4100, 0x0, 0x0)
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad:         /opt/gopath/src/github.com/hashicorp/nomad/client/driver/executor_plugin.go:97 +0x64
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: reflect.Value.call(0x104d900, 0x1286e98, 0x13, 0x1306748, 0x4, 0xc82007fee8, 0x3, 0x3, 0x0, 0x0, ...)
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad:         /opt/go/src/reflect/value.go:435 +0x120d
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: reflect.Value.Call(0x104d900, 0x1286e98, 0x13, 0xc82007fee8, 0x3, 0x3, 0x0, 0x0, 0x0)
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad:         /opt/go/src/reflect/value.go:303 +0xb1
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: net/rpc.(*service).call(0xc820174dc0, 0xc820174d80, 0xc820164980, 0xc82024c480, 0xc8201734e0, 0x113c640, 0xc820164c50, 0x199, 0xe8d700, 0xc8202f4100, ...)
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad:         /opt/go/src/net/rpc/server.go:383 +0x1c2
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad: created by net/rpc.(*Server).ServeCodec
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: nomad:         /opt/go/src/net/rpc/server.go:477 +0x49d
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [DEBUG] plugin: /usr/bin/nomad: plugin process exited
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [ERR] client: failed to start task 'whoami' for alloc '1003f732-e08d-7091-43a0-eda0c56bbc5d': error starting process via the plugin: unexpected EOF
Apr 29 14:38:26 <host> nomad[140535]: 2016/04/29 14:38:26 [INFO] client: Not restarting task: whoami for alloc: 1003f732-e08d-7091-43a0-eda0c56bbc5d
```

#### `nomad` config
```
addresses {
  http = "0.0.0.0"
}
bind_addr = "192.168.111.25"
data_dir = "/var/lib/nomad"
datacenter = "rnd"
disable_update_check = true
enable_syslog = true
log_level = "DEBUG"
name = "..."
syslog_facility = "LOCAL1"

client {
  enabled = true
  servers = ["192.168.111.11", "192.168.111.16", "192.168.111.21"]
  node_class = "dev"
  meta {
    node_count = "17"
  }
  network_interface = "docker0"
  network_speed = 10000
  options {
    consul.address = "192.168.111.25:8500"
    driver.raw_exec.enable = "true"
    docker.cleanup.container = "true"
    docker.cleanup.image = "false"
    docker.privileged.enabled = "true"
  }
  reserved {
    cpu = 200
    memory = 512
    disk = 1024
  }
}

telemetry {
  statsite_address = "127.0.0.1:8125"
}
```

Hopefully, this will help with tracking down the bug.

**Comment by diptanu** [comment]
@dvusboy I am not able to reproduce this with the same job file you have submitted. It's odd that things are crashing at that line.

**Comment by diptanu** [comment]
Also the version of server and client are same?

**Comment by dvusboy** [comment]
Yes

**Comment by dvusboy** [comment]
I'm replacing the executable at the endpoints with the published one (checksum: 28db8e8de24e1bbcd4da1494e9675a906e1bce50). I will rerun the test.

**Comment by dvusboy** [comment]
Replacing the executable with published one fix the segfault for the `whoami` test. Closing the issue.
